File: wrappers/javascript/shared/src/api/CredentialDefinition.ts
import type { Schema } from './Schema'

import { AnoncredsObject } from '../AnoncredsObject'
import { anoncreds } from '../register'

import { CredentialDefinitionPrivate } from './CredentialDefinitionPrivate'
import { KeyCorrectnessProof } from './KeyCorrectnessProof'

export type CreateCredentialDefinitionOptions = {
  originDid: string
  schema: Schema
  signatureType: string
  tag: string
  supportRevocation?: boolean
}

export class CredentialDefinition extends AnoncredsObject {
  public static create(options: CreateCredentialDefinitionOptions) {
    const { credentialDefinition, credentialDefinitionPrivate, keyProof } = anoncreds.createCredentialDefinition({
      originDid: options.originDid,
      schema: options.schema.handle,
      signatureType: options.signatureType,
      tag: options.tag,
      supportRevocation: options.supportRevocation ?? false,
    })

    return {
      credentialDefinition: new CredentialDefinition(credentialDefinition.handle),
      credentialDefinitionPrivate: new CredentialDefinitionPrivate(credentialDefinitionPrivate.handle),
      keyCorrectnessProof: new KeyCorrectnessProof(keyProof.handle),
    }
  }

  public static load(json: string) {
    return new CredentialDefinition(anoncreds.credentialDefinitionFromJson({ json }).handle)
  }

  public getId() {
    return anoncreds.credentialDefinitionGetAttribute({ objectHandle: this.handle, name: 'id' })
  }

  public getSchemaId() {
    return anoncreds.credentialDefinitionGetAttribute({ objectHandle: this.handle, name: 'schema_id' })
  }
}

File: wrappers/dotnet/lib/Models/CredentialDefinition.cs
using AnonCredsNet.Exceptions;
using AnonCredsNet.Interop;

namespace AnonCredsNet.Models;

public class CredentialDefinition : AnonCredsObject
{
    private CredentialDefinition(long handle)
        : base(handle) { }

    public static (
        CredentialDefinition CredDef,
        CredentialDefinitionPrivate CredDefPvt,
        KeyCorrectnessProof KeyProof
    ) Create(
        string schemaId,
        string issuerId,
        Schema schema,
        string tag,
        string sigType,
        string config
    )
    {
        if (
            string.IsNullOrEmpty(schemaId)
            || string.IsNullOrEmpty(issuerId)
            || schema == null
            || string.IsNullOrEmpty(tag)
            || string.IsNullOrEmpty(sigType)
            || string.IsNullOrEmpty(config)
        )
            throw new ArgumentNullException("Input parameters cannot be null or empty");

        // Parse config to determine if revocation should be supported
        var configObj = System.Text.Json.JsonSerializer.Deserialize<System.Text.Json.JsonElement>(
            config
        );
        var supportRevocation =
            configObj.TryGetProperty("support_revocation", out var revProp) && revProp.GetBoolean();

        var code = NativeMethods.anoncreds_create_credential_definition(
            schemaId,
            schema.Handle,
            tag,
            issuerId,
            sigType,
            supportRevocation,
            out var cd,
            out var pvt,
            out var proof
        );
        if (code != ErrorCode.Success)
            throw new AnonCredsException(code, AnonCreds.GetCurrentError());
        return (
            new CredentialDefinition(cd),
            new CredentialDefinitionPrivate(pvt),
            new KeyCorrectnessProof(proof)
        );
    }

    public static CredentialDefinition FromJson(string json) =>
        FromJson<CredentialDefinition>(json);
}

public class CredentialDefinitionPrivate : AnonCredsObject
{
    internal CredentialDefinitionPrivate(long handle)
        : base(handle) { }

    internal static CredentialDefinitionPrivate FromJson(string json) =>
        FromJson<CredentialDefinitionPrivate>(json);
}
